---
title: "QAA"
author: "Pranav Muthuraman"
date: "9/7/2022"
output: pdf_document
geometry: margin = 1.5cm
---
```{r setup, include=FALSE}
library(knitr)
library(formatR)
library(readr)
library(tidyverse)
opts_chunk$set(tidy.opts=list(width.cutoff=80),tidy=TRUE)

```
# Part 1
Per-base quality score plots were generated by running FASTQC version 11.5 on Talapas, using the following shell script:
```{bash, comment=NA}
cat fastqc_run.sh
```

## 8_2F_fox_S7_L008_R1_001.fastq.gz

**Quality Score Distribution for Fox Read 1**
```{r, out.width="300px"}
knitr::include_graphics("/Users/pranav_muthuraman/Desktop/QAA/fox_read1.png")
```

**Per base N Distribution for Fox Read 1**
```{r, out.width="300px"}
knitr::include_graphics("/Users/pranav_muthuraman/Desktop/QAA/fox_read1_N.png")
```

## 8_2F_fox_S7_L008_R2_001.fastq.gz
**Quality Score Distribution for Fox Read 2**
```{r, out.width="300px"}
knitr::include_graphics("/Users/pranav_muthuraman/Desktop/QAA/fox_read2.png")
```

**Per base N Distribution for Fox Read 2**
```{r, out.width="300px"}
knitr::include_graphics("/Users/pranav_muthuraman/Desktop/QAA/fox_read2_N.png")
```

## 14_3B_control_S10_L008_R1_001.fastq.gz
**Quality Score Distribution for Control Read 1**
```{r, out.width="300px"}
knitr::include_graphics("/Users/pranav_muthuraman/Desktop/QAA/control_read1.png")
```

**Per base N Distribution for Control Read 1**
```{r, out.width="300px"}
knitr::include_graphics("/Users/pranav_muthuraman/Desktop/QAA/control_read1_N.png")
```

## 14_3B_control_S10_L008_R2_001.fastq.gz
**Quality Score Distribution for Control Read 2**
```{r, out.width="300px"}
knitr::include_graphics("/Users/pranav_muthuraman/Desktop/QAA/control_read2.png")
```

**Per base N Distribution for Control Read 2**
```{r, out.width="300px"}
knitr::include_graphics("/Users/pranav_muthuraman/Desktop/QAA/control_read2_N.png")
```

### QUESTION 1
> For the per-base quality scores, the read1 files for both fox and control start a bit lower, then remain very high for the entire read. For the read2 files, fox starts a bit lower and has consistently lower quality throughout. The control starts even lower than the fox file, but improves to the same level after the 5th base pair. All 4 quality scores received a green passing mark. In all files, the N distributions are very low. 


I also produced these qualtity score plots using python; the script, the bioinfo module, and bash scripts used to run on talapas are shown here: 

Bioinfo module:
```{bash, comment = NA}
cat Bioinfo.py
```

Python script:
```{bash, comment = NA}
cat part1.py
```

Scripts run on Talapas:
```{bash, comment = NA}
cat 8_2F_fox_R1_qscore_dist_run.sh
```

```{bash, comment = NA}
cat 8_2F_fox_R2_qscore_dist_run.sh
```

```{bash, comment = NA}
cat 14_3B_control_R1_qscore_dist_run.sh
```

```{bash, comment = NA}
cat 14_3B_control_R2_qscore_dist_run.sh
```


Python Generated Plots:

**Python Gen: Per base N Distribution for Fox Read 1**
```{r, out.width="300px"}
knitr::include_graphics("/Users/pranav_muthuraman/Desktop/QAA/8_2F_fox_R1.png")
```

**Python Gen: Per base N Distribution for Fox Read 2**
```{r, out.width="300px"}
knitr::include_graphics("/Users/pranav_muthuraman/Desktop/QAA/8_2F_fox_R2.png")
```

**Python Gen: Per base N Distribution for Control Read 1**
```{r, out.width="300px"}
knitr::include_graphics("/Users/pranav_muthuraman/Desktop/QAA/14_3B_control_R1.png")
```

**Python Gen: Per base N Distribution for Control Read 2**
```{r, out.width="300px"}
knitr::include_graphics("/Users/pranav_muthuraman/Desktop/QAA/14_3B_control_R2.png")
```

### QUESTION 2
> All the distribution plots created by the python script look very similar to the ones created by FASTQC. The runtime to create each of these plots were: 21 mins, 21 mins, 3 mins, and 3 mins, respectively. The runtime of the entire fastQC was only 8 mins. FastQC may have run more quickly because it is able to read multiple files at once. 

### QUESTION 3
> Overall, the FOX files have better data quality than the CONTROL due to the quality scores. In both sets of data, the read 1 files have significantly better quality scores than the read 2 files.


# Part 2

### QUESTION 4
```
conda create --name QAA python=3.9
conda activate QAA
conda install -c bioconda cutadapt
conda install -c bioconda trimmomatic

cutadapt --version
4.1

trimmomatic -version
0.39
```

### QUESTION 5

cutadapt script:
```{bash, comment = NA}
cat cutadapt_run.sh
```

```
Cut adapt summary:
8_2F_fox:
Total read pairs processed:         36,482,601
  Read 1 with adapter:               2,145,600 (5.9%)
  Read 2 with adapter:               2,403,490 (6.6%)
Pairs written (passing filters):    36,482,601 (100.0%)

14_3B_control:
Total read pairs processed:          4,440,378
  Read 1 with adapter:                 264,208 (6.0%)
  Read 2 with adapter:                 299,716 (6.7%)
Pairs written (passing filters):     4,440,378 (100.0%)
```
```
SANITY CHECK:

forward adapters
zcat /projects/bgmp/shared/2017_sequencing/demultiplexed/8_2F_fox_S7_L008_R1_001.fastq.gz | 
grep 'AGATCGGAAGAGCACACGTCTGAACTCCAGTCA'
zcat /projects/bgmp/shared/2017_sequencing/demultiplexed/8_2F_fox_S7_L008_R2_001.fastq.gz | 
grep 'AGATCGGAAGAGCACACGTCTGAACTCCAGTCA' --> nothing
zcat /projects/bgmp/shared/2017_sequencing/demultiplexed/14_3B_control_S10_L008_R1_001.fastq.gz | 
grep 'AGATCGGAAGAGCACACGTCTGAACTCCAGTCA'
zcat /projects/bgmp/shared/2017_sequencing/demultiplexed/14_3B_control_S10_L008_R2_001.fastq.gz | 
grep 'AGATCGGAAGAGCACACGTCTGAACTCCAGTCA' --> nothing

reverse adapters
zcat /projects/bgmp/shared/2017_sequencing/demultiplexed/8_2F_fox_S7_L008_R1_001.fastq.gz | 
grep 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT' --> nothing
zcat /projects/bgmp/shared/2017_sequencing/demultiplexed/8_2F_fox_S7_L008_R2_001.fastq.gz | 
grep 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT' 
zcat /projects/bgmp/shared/2017_sequencing/demultiplexed/14_3B_control_S10_L008_R1_001.fastq.gz | 
grep 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT' --> nothing
zcat /projects/bgmp/shared/2017_sequencing/demultiplexed/14_3B_control_S10_L008_R2_001.fastq.gz | 
grep 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT' 
```
> Checked for forward primer in read1 and read2 files. Adapter in every read1 file, but not in read2 files. Checked for reverse primer in read1 and read2 files. Adapter in every read2 file, but not in read1 files.

## QUESTION 6

trimmomatic script
```{bash, comment = NA}
cat trimmomatic_run.sh
```

## QUESTION 7
To plot read length distributions, I ran FASTQC on the trimmed reads outputted from question 6. I only used the paired reads for these plots and all subsequent downstream analysis. 

trimmed FASTQC script
```{bash, comment = NA}
cat trimmed_fastqc_run.sh
```

```{r, include = F}

fox_r1_trim_len = read_tsv('trimmed_fastqc_output/8_2F_fox_S7_L008_R1_001.trimmed.paired_fastqc/fastqc_data.txt', n_max = 34)
colnames(fox_r1_trim_len) = c('Length','R1_Count')
fox_r2_trim_len = read_tsv('trimmed_fastqc_output/8_2F_fox_S7_L008_R2_001.trimmed.paired_fastqc/fastqc_data.txt', n_max = 34)
colnames(fox_r2_trim_len) = c('Length','R2_Count')

fox_trim_len = inner_join(fox_r1_trim_len,fox_r2_trim_len)

control_r1_trim_len = read_tsv('trimmed_fastqc_output/14_3B_control_S10_L008_R1_001.trimmed.paired_fastqc/fastqc_data.txt', n_max = 34)
colnames(control_r1_trim_len) = c('Length','R1_Count')
control_r2_trim_len = read_tsv('trimmed_fastqc_output/14_3B_control_S10_L008_R2_001.trimmed.paired_fastqc/fastqc_data.txt', n_max = 34)
colnames(control_r2_trim_len) = c('Length','R2_Count')

control_trim_len = inner_join(control_r1_trim_len,control_r2_trim_len)
```

```{r}
len_levels = c('34-35','36-37','38-39','40-41','42-43','44-45','46-47','48-49','50-51','52-53','54-55','56-57','58-59','60-61','62-63','64-65','66-67','68-69','70-71','72-73','74-75','76-77','78-79','80-81','82-83','84-85','86-87','88-89','90-91','92-93','94-95','96-97','98-99','100-101')
ggplot(data = fox_trim_len, mapping = aes(x = factor(Length, levels = len_levels))) + geom_point(mapping = aes(y = R1_Count, color = 'R1_Count')) + geom_point(mapping = aes(y = R2_Count, color = 'R2_Count')) + labs(title = 'Length Distribution in 8_2F_fox_S7 Reads After Trimming', x = 'Length', y = 'Number of Reads') + theme(axis.text.x=element_text(size = 9, angle = 90,vjust=0.3))
```
<br><br><br>
```{r}
ggplot(data = control_trim_len, mapping = aes(x = factor(Length, levels = len_levels))) + geom_point(mapping = aes(y = R1_Count, color = 'R1_Count')) + geom_point(mapping = aes(y = R2_Count, color = 'R2_Count')) + labs(title = 'Length Distribution in 14_3B_control_S10 Reads After Trimming', x = 'Length', y = 'Number of Reads') + theme(axis.text.x=element_text(size = 9, angle = 90,vjust=0.3))
```

> We can see that in both libraries, read 2 is  trimmed slightly more often than read 1. The adapter trimming rates should be about the same as they are dependent on insert length, and both reads of each library should have the same insert length. The small difference between read 1 and read 2 are probably due to the quality trimming rates, and we know that read 2 is of slighlty lower quality than read 1. 


# Part 3

### QUESTION 8

```
conda activate QAA
conda install -c bioconda star
conda install -c conda-forge numpy
conda install -c bioconda pysam
conda install -c conda-forge matplotlib
conda install -c bioconda htseq
```

### QUESTION 9

Generating alignment database script
```{bash, comment = NA}
cat star_makedb_run.sh
```

STAR alignment script
```{bash, comment = NA}
cat star_align_run.sh
```

### QUESTION 10

Python script from PS8 to report mapped/unmapped reads from SAM files
```{bash, comment = NA}
cat PS8.py
```

Output:
```
For 8_2F_fox_S7
Total: 69582314
Number of Mapped Reads: 69582314, Number of Unmapped Reads 0

For 14_3B_control_S10
Total: 8493304
Number of Mapped Reads: 8312390, Number of Unmapped Reads 180914

```

### QUESTION 11
```
conda install -c bioconda samtools
```

Script to sort SAM files before htseq
```{bash, comment = NA}
cat sort_sam_run.sh
```

Script to run htseq
```{bash, comment = NA}
cat htseq_count_run.sh
```


### QUESTION 12

Script to analyze htseq output for strand specificity
```{bash, comment = NA}
cat analyze_htseq_count_run.sh
```


Output:
```
8_2F_fox_S7_L008.stranded.genecount:
reads mapping to feature:
1282235
total reads:
34791157

8_2F_fox_S7_L008.reverse.genecount:
reads mapping to feature:
28041293
total reads:
34791157

14_3B_control_S10_L008.stranded.genecount:
reads mapping to feature:
167859
total reads:
4246652

14_3B_control_S10_L008.reverse.genecount:
reads mapping to feature:
3667140
total reads:
4246652
```

> I can propose that these libraries are indeed strand specific as 80% of the fox reads mapped to features on the reverse strand and 86% of the control reads mapped to features on the reverse strand. This tells us that a majority of the reads are specific to the reverse strand. If this was an unstranded kit, we would have roughly the same amount on both the forward and reverse strand. 













